File: next/dweb-browser/src/Platforms/iOS/MicroModule/Core/WindowBounds.cs
using System.Text.Json.Serialization;

namespace DwebBrowser.MicroService.Core;

/// <summary>
/// 窗口大小与位置
///
/// 默认值是NaN，这种情况下，窗口构建者需要自己对其进行赋值
/// </summary>
/// <param name="left"></param>
/// <param name="top"></param>
/// <param name="width"></param>
/// <param name="height"></param>
public record WindowBounds(float left = float.NaN, float top = float.NaN, float width = float.NaN, float height = float.NaN)
{
    public class Mutable : IEquatable<Mutable>
    {
        [JsonPropertyName("left")]
        public float Left { get; set; }

        [JsonPropertyName("top")]
        public float Top { get; set; }

        [JsonPropertyName("width")]
        public float Width { get; set; }

        [JsonPropertyName("height")]
        public float Height { get; set; }

        public Mutable(float left = float.NaN, float top = float.NaN, float width = float.NaN, float height = float.NaN)
        {
            Left = left;
            Top = top;
            Width = width;
            Height = height;
        }

        public bool Equals(Mutable other) => GetHashCode() == other.GetHashCode();

        public override int GetHashCode()
        {
            return Left.GetHashCode() ^
                Top.GetHashCode() ^
                Width.GetHashCode() ^
                Height.GetHashCode();
        }

        public WindowBounds ToImmutable() => new(Left, Top, Width, Height);
    }

    public static readonly WindowBounds Zero = new(0f, 0f, 0f, 0f);

    public Mutable ToMutable() => new(left, top, width, height);
}

File: next/kmp/window/src/commonMain/kotlin/org/dweb_browser/sys/window/core/WindowBounds.kt
package org.dweb_browser.sys.window.core

import kotlinx.serialization.Serializable

/**
 * 窗口大小与位置
 *
 * 默认值是NaN，这种情况下，窗口构建者需要自己对其进行赋值
 */
@Serializable
data class WindowBounds(
  val left: Float = Float.NaN,
  val top: Float = Float.NaN,
  val width: Float = Float.NaN,
  val height: Float = Float.NaN,
) {
  companion object {
    val Zero = WindowBounds(0f, 0f, 0f, 0f)
  }

  fun toMutable() = Mutable(left, top, width, height)
  class Mutable(
    var left: Float,
    var top: Float,
    var width: Float,
    var height: Float,
  ) {
    fun toImmutable() = WindowBounds(left, top, width, height)
  }
}